test/junit/scala/collection/immutable/ListSetTest.scala

```diff
@@ -50,4 +52,32 @@ class ListSetTest {
     val s = ListSet(1, 2, 3, 5, 4)
     assertEquals(List(1, 2, 3, 5, 4), s.iterator.toList)
   }
+
+  @Test
+  def hasCorrectOrderAfterPlusPlus(): Unit = {
+    val foo = ListSet(1)
```

Comment:
I think we should exercise at least lengths 0 - 5 to make sure we cover the cases hit by the code.

Although, if scala check is permitted in this repo, I would definitely like to see a scala check like:

```
forAll { (l1: ListSet[Int], l2: ListSet[Int]) =>
  (l1 ++ l2).iterator.toList == (l1.iterator.toList ++ (l2.iterator.filterNot(l1).toList))
}
```
I think that's the property we are checking here.

Even without scala check, you could test that manually:
```scala
def property(l1: ListSet[Int], l2: ListSet[Int]) = {
  assertEquals(l1 ++ l2).iterator.toList, l1.iterator.toList ++ (l2.iterator.filterNot(l1).toList))
}

for {
  start0 <- (0 until 6)
  end0 <- (start until 6)
  start1 <- (0 until 6)
  end1 <- (start until 6)
  ls0 = ListSet((start0 until end0) : _*)
  ls1 = ListSet((start1 until end1) : _*)
} property(ls0, ls1)
```

That should expose the bug we have currently.

Reply:
ScalaCheck tests are very much encouraged here.